### DISCUSSION TITLE
Question about restoring official LEGO firmware

### DISCUSSION THREAD

**Discussion by marc-hg** [pybricks/support#1225] [Q&A]
When I try to restore the official LEGO firmware once it ends it doesn't work out of the box in the official spike legacy webapp or desktop app (I have to upgrade to 3.0, then downgrade to 2.0 then for some reason upgrade again in the lego legacy app)

Is there any way to go from pybricks firmware to the official spike legacy firmware from pybricks? The one used at https://spikelegacy.legoeducation.com/

**Comment by laurensvalk** [comment]
Thanks for reporting.

> it doesn't work out of the box in the official spike legacy webapp

This is supposed to work, although it is supposed to give you an update to the latest V2 if available.

Is this not happening?

**Reply by marc-hg** [reply]
No, from what I've tried it seems that something happens once you upgrade to official lego 3 firmware that "corrupts" the robot in some way.

Steps to reproduce:

1. Upgrade to 3 official firmware: https://spike.legoeducation.com/
2. Downgrade to legacy 2 firmware in pybricks web
3. Try to connect in legacy spike page https://spikelegacy.legoeducation.com/ (it will fail to connect and say a hub os update is required)

**Reply by laurensvalk** [reply]
> Try to connect in legacy spike page https://spikelegacy.legoeducation.com/ (it will fail to connect and say a hub os update is required)

And when you try this ("hub os update is required") does it work?

**Reply by marc-hg** [reply]
Yes

**Reply by marc-hg** [reply]
What's more interesting is that after doing the hub os update, if I go back to pybricks firmware and then back to spike legacy (using pybricks installer) it works (until I go back to official firmware 3)

**Reply by laurensvalk** [reply]
I just checked, and both Pybricks recovery and the LEGO Education downgrader install: MicroPython v1.14-876-gfbecba865 on 2021-05-04, so I'm not sure what the difference is.

> > And when you try this ("hub os update is required") does it work?
> 
> Yes

This sounds good though, so maybe it is acceptable since our recovery tool states.

> Once the recovery is complete, use the LEGO® Education™ SPIKE™ app to get the latest official firmware. You can use version 2 (legacy) or version 3 to do so.

**Reply by dlech** [reply]
> so I'm not sure what the difference is.

I think the difference is the file system on the external flash. If the hub previously had the LEGO V3 file system, then when downgrading, the external flash would need to be wiped and reinstalled by the legacy LEGO app.

**Reply by laurensvalk** [reply]
Right, that's why an update is always needed as stated in our notes.

But the restored firmware should at least wipe the file system, so going back and forth as OP describes (V3->Pybricks->V2->V3->V2) should not be required.

Pybricks->V2 should work immediately after the auto-update when connecting it to the V2 app.

**Reply by marc-hg** [reply]
I can confirm V3 -> Pybricks -> V2 -> auto-update works

I'm still not quite sure why after that V2 -> Pybricks -> V2 works without the auto-update

Is there any way to get V3 -> Pybricks -> V2 working without the auto-update?

**Reply by laurensvalk** [reply]
I don't think so --- I think the official downgrader works essentially the same as ours. It downgrades the firmware, but leaves it up to the app to restore the "runtime" files like sounds and the on-screen menu, which get wiped when installing V3.

> I'm still not quite sure why after that V2 -> Pybricks -> V2 works without the auto-update

Because we saw this coming, Pybricks does not touch the file system at all :smile: So when you come from V2 and go back to V2 later, all your files are still there.

**Reply by laurensvalk** [reply]
Out of curiosity, what is the use case of using V2 and V3 together?

Or conversely, what could we do to make Pybricks more useful by itself, so that switching back was not needed? :)

**Reply by marc-hg** [reply]
On some occasions I require usb connectivity and V2 works with that so I use both pybricks and V2.

I actually don't use V3 for anything but it's not as convenient to downgrade to V2 as it is to change it to pybricks since it becomes a two step process.

Thanks a lot for the explanation though.

**Reply by laurensvalk** [reply]
So just to reiterate, V2 -> Pybricks -> V2 is expected to work without additional updates.

I'll have to try this myself to see what the issue might be.

**Reply by marc-hg** [reply]
Yeah,  V2(working V2) -> Pybricks -> V2 works without additional updates from what I've tested. The issue only comes up when you upgrade to V3 (I assume because of the filesystem)

**Comment by smolto** [comment]
@dlech, if difference is the file system on the external flash, Do you think it would be feasible to implement what LEGO does in their app? Because at the end of 2023 I understand that this process will be deprecated and downgrading will be more complex.

**Reply by dlech** [reply]
What is being deprecated and where did you hear it?

Since LEGO has fairly frequent updates, we didn't think it makes sense to try to keep up with that. It seems easier to just use the LEGO apps for that part to ensure it is always what is needed by the version of the LEGO app that is installed by the user.

**Reply by smolto** [reply]
I'm sorry if I explained myself wrong, I don't know what they are going to depreciate or when, I was just assuming that scenario to ask you the question of the feasibility of developing that feature. Perfect, from what you have been able to tell me, it is best to use LEGO applications for these processes because they will always install what is necessary to work.

**Reply by Kikketer** [reply]
I'm having a similar issue with the general idea that "implement what LEGO does in their app".  Right now I have a Mindstorms (blue controller) and I'd like to convert it to a Spike Prime V3.

So I "restore" in pybricks claiming that it's the Spike controller (yellow) which works fine to convert the mindstorms controller to spike.  But it's updated with V2 of spike, so when I go to the LEGO app I need to upgrade to V3.  But the lego app throws an error when I attempt that.  I assume there's something different or detected that it's mindstorms vs spike.

Path: Mindstorms -> PyBricks restore to Spike (yellow) -> LEGO App Upgrade -> error thrown.

**Reply by laurensvalk** [reply]
Since V3, the LEGO Education app does not support/allow using the MINDSTORMS hub. This is not a technical limitation, so there might be other reasons. (It would have been nice if their error just said that, though.)

Likewise, we don't know if we'd have permission to offer that path, so we don't.

That said, we have a really exciting update in the pipeline which might mean you wouldn't even need that path anymore. (Hint: Think Pybricks, but you don't have to write any code... :wink:)

**Reply by Kikketer** [reply]
Awesome thank you @laurensvalk I figured if anyone knew the answer it would be the folks behind PyBricks.
I'll wait impatiently for this sweet update in the pipe!

I'll hang out in PyBricks land with my older controller.  Thanks for keeping this thing relevant.
